dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java

```diff
@@ -77,6 +84,40 @@ public String toString() {
                 + cf.channelName() + ", delegate=" + delegate + '}';
     }
 
+    static final class StateHolderKey<T> {
+        private final Class<T> valueClass;
+        private final Supplier<T> factory;
+
+        T cast(final Object value) {
+            return valueClass.cast(value);
+        }
+
+        Supplier<T> getFactory() {
+            return factory;
+        }
+
+        StateHolderKey(final Class<T> valueClass, Supplier<T> factory) {
+            this.valueClass = valueClass;
+            this.factory = factory;
+        }
+    }
+
+    static final class StateHolder {
```

Comment:
disclaimer: naming is hard, and I'm not very good at it. Take my comments with a grain of salt!

Thoughts on extracting this to its own file, and renaming to `ChannelState`? "Holder" makes me think it's using the old [initialize-on-demand holder pattern](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom).

If we extract to its own file, the key type could make sense as a static nested class, since the outer class provides context as to what it's a key for e.g. `ChannelState.Key`.

Reply:
yep done